### ISSUE TITLE
Cascaded/overridden app-wide themes are unnecessarily blocked by ag-Grid

### ISSUE THREAD

**Issue by etramell-tem** [ag-grid/ag-grid#4342]
I learned about ag-Grid at the November 2020 Enterprise NG conference (thanks, @blove!), and I'm currently tasked with using ag-Grid to replace our Angular app's painful `<mat-table>` and `<cdk-virtual-scroll-viewport>` Franken-solution.

So far so good!  👍 

Except for one particular theming issue...

**I'm submitting a ...**  (check one with "x")
```
[x] bug report => see 'Providing a Reproducible Scenario'
[] feature request => do not use Github for feature requests, see 'Customers of ag-Grid'
[?] support request => see 'Requesting Community Support'
```
**Current behavior**

Our app uses Angular Material.  We need standard-looking Material tables (`ag-theme-material`) on most pages, but modified-looking versions (e.g. tighter spacing, smaller font, etc.) for specific pages.

We're using the `<ag-grid-angular>` component to place tables on our pages.

We have this mixin to style tables for most of our app:
```scss
@import '~ag-grid-community/src/styles/ag-grid.scss';
@import '~ag-grid-community/src/styles/ag-theme-material/sass/ag-theme-material-mixin';

@mixin ag-theme-OUR-APP($theme, $theme-override-parameters: ()) {
  $background: map-get($theme, background);
  $foreground: map-get($theme, foreground);
  $accent: map-get($theme, accent);

  ag-grid-angular {
    @include ag-theme-material(
      map-merge(
        (
          foreground-color: mat-color($foreground, base),
          background-color: mat-color($background, background),
          header-background-color: ag-derived(background-color),
          row-hover-color: mat-color($foreground, base, 0.1),
          selected-row-background-color: mat-color($accent, 0.3),
          border-color: mat-color($foreground, base, 0.2)
        ),
        $theme-override-parameters
      )
    );
  }
}
```
... where `OUR-APP` is actually our app's name.

(It probably looks weird to see `ag-grid-angular` in there as a selector instead of `.ag-theme-material` as it is in most examples, but that just lets us avoid adding a CSS class since we always use that component.)

This works great to give us a consistent app-wide Material theme for our tables.  (BTW, along with some other SCSS configuration, this also enables us to support dark mode for Material-themed ag-Grid tables, but that's beyond the scope of this bug report.)

Now, we get to a component that needs a variation on that theme, and we have something like this, for example:
```scss
some-specific-component {
  @include ag-theme-OUR-APP($theme, (
    foreground-color: purple,
    font-size: 12px
    // whatever other theme overrides we want for this component
  ));
}
```
This makes lots of sense, except that when we try to build our stylesheets, a hardcoded error in ag-Grid's mixins stops the build with an error like this:
```
Theme CSS for SOME_CSS_SELECTOR has been generated twice. This can happen when you @import ag-theme-xyz.scss
twice, or @import the legacy ag-theme-xyz.scss and then also @include the ag-theme-xyz() mixin. If you're
using the mixin, just @import ag-theme-xyz-mixin.scss.
```
... where SOME_CSS_SELECTOR is a selector specific to our app.

Note that:
- we are not using a legacy theme.
- we are, however, including `ag-theme-material` `N` times, where `N = 1 + the number of components that override our app's  theme`.  (Once directly, and `N - 1` times indirectly.)

I tracked this down in `node_modules/ag-grid-community/src/styles/mixins/_ag-theme-mixins.scss`:
```scss
$-ag-included-themes: () !default;

@mixin ag-include-theme-once-only() {
    $name: nth(nth(&, 1), 1); // returns name of top-level selector, e.g. ".ag-theme-balham"
    @if index($-ag-included-themes, $name) != null {
        @error "Theme CSS for #{$name} has been generated twice. This can happen when you @import ag-theme-xyz.scss twice, or @import the legacy ag-theme-xyz.scss and then also @include the ag-theme-xyz() mixin. If you're using the mixin, just @import ag-theme-xyz-mixin.scss.";
    }
    $-ag-included-themes: append($-ag-included-themes, $name) !global;
}
```
**Expected behavior**

I expected to be able to take advantage of standard SCSS mixins and CSS cascades/overrides without getting an error message.

From the wording of the error message, it makes sense to me why this check was added -- especially for those migrating from earlier versions of ag-Grid to newer ones.  But on the other hand, it completely stops me in my tracks and prevents me from doing something I should be able to do if I deem it necessary.

**Workaround**

I wasted (or "invested", if you prefer) a lot of time figuring out what was causing this and how to avoid it.  I assumed for way too long that I was doing something wrong (since I'm just learning ag-Grid), but I finally concluded that I wasn't.

Luckily, I ended up with a very simple workaround.  I added this above our app's `@mixin ag-theme-OUR-APP`:
```scss
/*
  Hack to make ag-Grid allow theme overrides (for standard app-wide table theme vs. page-specific).

  ag-Grid helpfully protects apps from generating unnecessary CSS when a developer accidentally includes one of
  their themes more than once.  But that also prevents apps from including them twice on purpose for use cases
  like ours, so we disable that check by rewriting their mixin with an empty body.

  Original mixin location: node_modules/ag-grid-community/src/styles/mixins/_ag-theme-mixins.scss
*/
@mixin ag-include-theme-once-only() {}
```
And now we get all the theming and behavior we want with no SCSS build errors.

**Question**

Is there is a better existing solution than my workaround/hack?  I sure don't see one, but the documentation is massive, so maybe I missed something.

**Proposed fix**

If the ag-Grid team thinks this is worthy of a fix for the community version, I'd think it would be pretty simple and desirable to add a new property to the `GridOptions` interface, named `suppressThemeInclusionMoreThanOnceError` (or something shorter that says the same thing :wink: ).  Of course, appropriate "don't do this unless you know what you're doing" documentation would probably make sense for this.

**Please tell us about your environment:**

* **ag-Grid version:** 25.0.1

* **Browser:** irrelevant

* **Language:** modern Angular (not AngularJS)

**Thanks!**
If you read this far, I appreciate your attention to this bug report.  But if you decide to just go ahead and close it, at least I've provided some information to future Googlers trying to figure out this issue.  Hopefully this will save someone else some time and frustration!

Keep up the good work, ag-Grid Team!  🥇

**Comment by blove** [comment]
Thank @etramell-tem for the shout out. 😎

I suspect that the potential issue that the ag-Grid team is trying to avoid for you here is importing the (likely massive) amount of CSS for the grid more than once. So, they've implemented a safeguard for you. The best solution for you in this scenario is likely to use CSS class selectors to target the specific aspects of the grid that you need to override the style of. I recognize this is probably more effort, but would minimize your bundle size.

**Comment by etramell-tem** [comment]
@blove - Thanks for your quick response!  Totally makes sense.

We did discuss the fact that we are probably generating too much output with our workaround (most of which the browser will have to parse and eventually throw away for the specific components that do the overriding).

But we're weighing that against the fact that using class selectors instead of theme properties would require us to override lots of settings, depending on how many other properties are generated using `ag-derived()` with the property in question.  (I assume that's the "more effort" you are referring to.)

Hopefully we won't be doing this overriding of our app's standard theme very much.  But if it does turn out that we've really hurt our app's bundle size this way, then yes, we'll be forced to do the more brute-force (and more error-prone and less future-proof) thing.  But -- make it work first and optimize later, right?  🙂 

I do think the safeguard is a great idea, hence the suggestion to allow suppression if needed, not totally remove it.

**Comment by makinggoodsoftware** [comment]
@etramell-tem 

Do you think you could provide us with a working repo where we can see this error? If you do so, I will raise here internally to see if we can see if there is a workaround

Would that be an option?

@blove Thanks for the help :) And for the best ag-grid courses! Nice to see you out here

**Comment by etramell-tem** [comment]
@makinggoodsoftware - Yes, I can try to make a repo for you, but I need to finish some other stuff first.

How long do I have before this issue will expire?

**Comment by makinggoodsoftware** [comment]
Hi,

There is no rush, you could get back to us in a few days from now, I'll try to keep this open to give you time to get back

Thanks

**Comment by etramell-tem** [comment]
@makinggoodsoftware - Just checking in so you know I haven't forgotten about this...

I'm almost done with the ag-Grid-ification (sorry, AG Grid-ification) of the tables on my site, after which I should have a little time to make the example app that shows this problem.

Thanks!

**Comment by makinggoodsoftware** [comment]
Hi,

Thanks for the update! 

Will keep this open!

**Comment by makinggoodsoftware** [comment]
Hi,

Going to close this one, but if you do manage to reproduce it, by all means, feel free to open a new issue, make sure to link this one so we can keep the history

Hope this makes sense!

Thanks

**Comment by etramell-tem** [comment]
@makinggoodsoftware - This completely makes sense, and I apologize for not getting back to it.

The workaround is working fine for us, but I still have this on my todo list because I feel I can give back in a small way and perhaps help other users by giving you this example.  I will create a new issue when I'm ready, and I'll definitely link back here.
